### PR TITLE
feat(scan): scan time.Time sets the default decoding

### PR DESCRIPTION
**Pull request by monkey92t**
1. scan time.Time, we should set the default decoding format, because we encoded time.Time in RFC3339Nano format in proto.WriteArg.
```go
case time.Time:
		w.numBuf = v.AppendFormat(w.numBuf[:0], time.RFC3339Nano)
		return w.bytes(w.numBuf)
```
We should achieve the consistency of Write(Hset Scan) and Read(HMGet Scan):

```go
type data struct {
	Key1 string    `redis:"key1"`
	Key2 int       `redis:"key2"`
	Time time.Time `redis:"time"`
}

// write set
client.HSet(ctx, "hash", &data{
	Key1: "hello",
	Key2: 100,
	Time: time.Now(),
})

// read get
client.HMGet(ctx, "hash", "key1", "key2", "time").Scan(&data)
```

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

2. `switch case encoding.BinaryMarshaler` is invalid

### CODE REVIEW

**Comment by vmihailenco** [comment]
>switch case encoding.BinaryMarshaler is invalid

Could you elaborate? I don't see why it should not work...

**Review by leoantony72 — DISMISSED** [review]

**Comment by monkey92t** [comment]
> > switch case encoding.BinaryMarshaler is invalid
> 
> Could you elaborate? I don't see why it should not work...

Oh..
it my test error, my mistake, I recover it.

**Comment by leoantony72** [comment]
hgetall with Scan command doesn't work if there is time.Time field in the struct and it leads to erratic behaviour like not retrieving every field from the hashset

**Comment by monkey92t** [comment]
We can directly assert that the `encoding.TextUnmarshaler` interface, `time.Time` implements the `encoding.TextUnmarshaler` interface, which is more convenient. Consider removing `Scanner` in the future.

**Comment by leoantony72** [comment]
> 

That would remove the use of Scan. Which is more reliant

File: internal/hscan/structmap.go [hunk 1]
```diff
@@ -97,7 +100,10 @@ func (s StructValue) Scan(key string, value string) error {
 	}
 
 	if isPtr && v.Type().NumMethod() > 0 && v.CanInterface() {
-		if scan, ok := v.Interface().(Scanner); ok {
+		switch scan := v.Interface().(type) {
+		case encoding.TextUnmarshaler:
+			return scan.UnmarshalText(util.StringToBytes(value))
+		case Scanner:
```
Comment (vmihailenco) [inline internal/hscan/structmap.go:106]: Perhaps swap Scanner and TextUnmarshaler? Scanner looks more specific.

**Review by vmihailenco — DISMISSED** [review]

File: internal/hscan/structmap.go (hunk 1 above)
Comment (leoantony72) [reply internal/hscan/structmap.go:106]: ah

**Review by chayim — APPROVED** [review]

**Comment by ryan961** [comment]
> We can directly assert that the `encoding.TextUnmarshaler` interface, `time.Time` implements the `encoding.TextUnmarshaler` interface, which is more convenient. Consider removing `Scanner` in the future.

hset struct field can assert the `encoding.TextMarshaler` interface ?
https://github.com/redis/go-redis/issues/2161#issuecomment-1471422636

**Pull request merged**